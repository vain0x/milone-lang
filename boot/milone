#!/bin/bash

dotnet run -p MiloneLang -c Release "$@"
