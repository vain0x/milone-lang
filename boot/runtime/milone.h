// Runtime code for C programs generated by the milone-lang compiler.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct String {
  char* str;
  int len;
};

int int_cmp(int l, int r) {
  if (l == r) return 0;
  if (l < r) return -1;
  return 1;
}

int int_clamp(int x, int l, int r) {
  if (x < l) return l;
  if (x > r) return r;
  return x;
}

int str_cmp(struct String left, struct String right) {
  return strcmp(left.str, right.str);
}

struct String str_add(struct String left, struct String right) {
  if (left.len == 0 || right.len == 0) {
    return right.len == 0 ? left : right;
  }
  int len = left.len + right.len;
  char* str = (char*)malloc((len + 1) * sizeof(char));
  strcpy(str, left.str);
  strcpy(str + left.len, right.str);
  return (struct String){.str = str, .len = len};
}

struct String str_substring(int l, int r, struct String s) {
  l = int_clamp(l, 0, s.len);
  r = int_clamp(r + 1, l, s.len);
  int len = r - l;
  char* str;
  if (r == s.len) {
    str = s.str + l;
  } else {
    str = (char*)malloc((len + 1) * sizeof(char));
    strncpy(str, s.str + l, len);
    str[len] = '\0';
  }
  return (struct String){.str = str, .len = len};
}

int str_to_int(struct String s) {
  return atoi(s.str);
}

struct String str_of_int(int value) {
  char* str = (char *)calloc(20, sizeof(char));
  sprintf(str, "%d", value);
  return (struct String){.str = str, .len = strlen(str)};
}

struct String str_of_char(char value) {
  char* str = (char *)calloc(2, sizeof(char));
  str[0] = value;
  return (struct String){.str = str, .len = strlen(str)};
}

void milone_assert(int cond) {
  if (!cond) {
    exit(1);
  }
}

struct String file_read_all_text(struct String file_name) {
  FILE *fp = fopen(file_name.str, "r");
  if (!fp) {
    fprintf(stderr, "File '%s' not found.", file_name.str);
    abort();
  }

  fseek(fp, 0, SEEK_END);
  long size = ftell(fp);
  if (size < 0) {
    fclose(fp);
    fprintf(stderr, "%s", "Couldn't retrieve the file size.");
    abort();
  }
  fseek(fp, 0, SEEK_SET);

  char *content = (char *)calloc((size_t)size + 1, sizeof(char));
  size_t read_size = fread(content, 1, (size_t)size, fp);
  if (read_size != (size_t)size) {
    fclose(fp);
    fprintf(stderr, "%s", "Couldn't retrieve the file contents");
    abort();
  }

  fclose(fp);
  return (struct String){.str = content, .len = size};
}

int file_write_all_text(struct String file_name, struct String content) {
  FILE *fp = fopen(file_name.str, "w");
  if (!fp) {
    fprintf(stderr, "File '%s' not found.", file_name.str);
    abort();
  }

  fprintf(fp, "%s", content.str);

  fclose(fp);
  return 0;
}

static int s_argc;
static char **s_argv;

int arg_count(int _unit) {
  return s_argc;
}

struct String arg_get(int index) {
  if (!(0 <= index && index < s_argc)) {
    abort();
  }

  char *str = s_argv[index];
  int len = strlen(str);
  return (struct String){.str = str, .len = len};
}

int main(int argc, char** argv) {
  s_argc = argc - 1;
  s_argv = argv + 1;

  int milone_main();
  return milone_main();
}

#define main milone_main
