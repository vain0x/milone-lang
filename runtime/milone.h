// Runtime code for C programs generated by the milone-lang compiler.
// Interface is still unstable.

#ifndef MILONE_H_INCLUDED
#define MILONE_H_INCLUDED

#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

// -----------------------------------------------
// assertion
// -----------------------------------------------

void milone_assert(bool cond, int row, int column);

// -----------------------------------------------
// memory management
// -----------------------------------------------

void milone_enter_region(void);
void milone_leave_region(void);
void *milone_mem_alloc(int count, size_t size);

// -----------------------------------------------
// int
// -----------------------------------------------

int int_compare(int l, int r);
int int64_compare(int64_t l, int64_t r);
int uint64_compare(uint64_t l, uint64_t r);

// -----------------------------------------------
// str
// -----------------------------------------------

// This assumes some invariants, see documentation about string type.
struct String {
    char const *str;
    int len;
};

// Convert a null-terminated string to a string object by computing its length.
//
// SAFETY: The resulting string is valid until the pointer is valid.
struct String str_borrow(char const *c_str);

// Compare two strings in lexicographical order.
int str_compare(struct String l, struct String r);

// Create a copy of native C string.
struct String str_of_raw_parts(char const *ptr, int len);

// Create a copy of native null-terminated C string.
struct String str_of_c_str(char const *s);

// Create a concatenation of two string.
struct String str_add(struct String l, struct String r);

// Create a slice of string.
struct String str_get_slice(int l, int r, struct String s);

struct String str_ensure_null_terminated(struct String s);

// Ensure null-terminated.
char const *str_to_c_str(struct String s);

// Conversion:

int8_t str_to_int8(struct String s);
int16_t str_to_int16(struct String s);
int str_to_int(struct String s);
int64_t str_to_int64(struct String s);
intptr_t str_to_intptr(struct String s);

uint8_t str_to_uint8(struct String s);
uint16_t str_to_uint16(struct String s);
uint32_t str_to_uint32(struct String s);
uint64_t str_to_uint64(struct String s);
uintptr_t str_to_uintptr(struct String s);

struct String str_of_int64(int64_t value);
struct String str_of_uint64(uint64_t value);

double str_to_double(struct String s);
struct String str_of_double(double value);

char str_to_char(struct String s);
struct String str_of_char(char value);

// Actual name of string list.
struct StringList;

struct String str_concat(struct String sep, struct StringList const *strings);

// Low level operations.

inline char const *str_as_ptr(struct String s) { return s.str; }

#endif
