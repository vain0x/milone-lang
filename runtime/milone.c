// Runtime code for C programs generated by the milone-lang compiler.

// Some of functions are not declared in milone.h
// but usable with __nativeFun.

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <milone.h>

// -----------------------------------------------
// memory management (memory pool)
// -----------------------------------------------

// structure for memory management. poor implementation. thread unsafe.
struct MemoryChunk {
    // start of allocated memory or null
    void *ptr;

    // used size
    size_t len;

    // total capacity
    size_t cap;

    // list of used chunks or null
    // each chunk is allocated by calloc and should be freed after leaving
    // region
    struct MemoryChunk *used;

    // parent region or null
    // each region is allocated by calloc and should be freed after leaving
    // region
    struct MemoryChunk *parent;
};

static struct MemoryChunk s_heap;
static int s_heap_level; // depth of current region
static int s_heap_size;  // consumed size in all regions
static int s_heap_alloc; // allocated size in all regions

static void oom(void) {
    fprintf(stderr, "Out of memory.\n");
    exit(1);
}

// free chunk itself, not used list
static void free_chunk(struct MemoryChunk *chunk) {
    assert(chunk != NULL);
    assert(chunk->ptr != NULL);

    free(chunk->ptr);
    s_heap_size -= chunk->len;
    s_heap_alloc -= chunk->cap;
}

void milone_enter_region(void) {
    // fprintf(stderr, "debug: enter_region level=%d size=%d\n", s_heap_level +
    // 1, s_heap_size);

    struct MemoryChunk *parent = calloc(1, sizeof(struct MemoryChunk));
    if (parent == NULL) {
        oom();
    }

    *parent = s_heap;
    s_heap = (struct MemoryChunk){.parent = parent};

    s_heap_level++;
}

void milone_leave_region(void) {
    // fprintf(stderr, "debug: leave_region level=%d size=%d\n", s_heap_level,
    // s_heap_size);

    assert(s_heap.parent != NULL);
    struct MemoryChunk *parent = s_heap.parent;

    // free current region
    {
        if (s_heap.ptr != NULL) {
            free_chunk(&s_heap);
        }

        // other than s_heap, chunk itself needs to be freed
        struct MemoryChunk *chunk = s_heap.used;
        while (chunk) {
            struct MemoryChunk *current = chunk;
            chunk = current->used;

            free_chunk(current);
            free(current);
        }

        s_heap = *parent;
        free(parent);
    }

    assert(s_heap_level > 0);
    s_heap_level--;

    // fprintf(stderr, "debug: free level=%d size=%d\n", s_heap_level,
    // s_heap_size);
}

// allocate new chunk
static void *milone_mem_alloc_slow(size_t total) {
    size_t cap = s_heap.cap;

    // move current chunk to used list
    if (s_heap.ptr != NULL) {
        struct MemoryChunk *used = calloc(1, sizeof(struct MemoryChunk));
        if (used == NULL) {
            oom();
        }

        *used = s_heap;
        s_heap.used = used;
    }

    // grow exponentially
    cap *= 2;
    if (cap < total) {
        cap = total;
    }

    // initialize chunk
    void *ptr = calloc(1, cap);
    if (ptr == NULL) {
        oom();
    }
    s_heap.ptr = ptr;
    s_heap.len = total;
    s_heap.cap = cap;

    s_heap_size += total;
    s_heap_alloc += cap;

    // fprintf(stderr, "debug: alloc level=%d size=%d  new chunk(h=%d
    // cap=%d)\n", s_heap_level, (int)size, h, (int)cap);
    return ptr;
}

void *milone_mem_alloc(int count, size_t size) {
    assert(count > 0 && size > 0);

    size_t total = (size_t)count * size;

    // align to 16x
    total = ((total & ~15) | 15) + 1;
    assert(total % 16 == 0);

    if (s_heap.len + total > s_heap.cap) {
        return milone_mem_alloc_slow(total);
    }

    // use current chunk
    void *ptr = (char *)s_heap.ptr + s_heap.len;
    s_heap.len += total;
    s_heap_size += total;

    // fprintf(stderr, "debug: alloc level=%d size=%d (%dx%d)\n", s_heap_level,
    // (int)((size_t)count * size), (int)count, (int)size);
    return ptr;
}

// -----------------------------------------------
// int
// -----------------------------------------------

int int_compare(int l, int r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int int64_compare(int64_t l, int64_t r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int uint64_compare(uint64_t l, uint64_t r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int int_clamp(int x, int l, int r) {
    if (x < l)
        return l;
    if (x > r)
        return r;
    return x;
}

// -----------------------------------------------
// str
// -----------------------------------------------

struct StringBuilder {
    char *buf;
    int len;
    int cap;
};

static struct StringBuilder *string_builder_new_with_capacity(int cap) {
    assert(cap >= 0);

    struct StringBuilder *sb =
        milone_mem_alloc(1, sizeof(struct StringBuilder));
    *sb = (struct StringBuilder){
        .buf = cap > 0 ? milone_mem_alloc(cap, sizeof(char)) : "",
        .len = 0,
        .cap = cap,
    };
    return sb;
}

static void string_builder_grow(struct StringBuilder *sb, int addition) {
    assert(addition >= 0);
    assert(sb->len + addition >= sb->cap);

    // grow exponentially
    sb->cap *= 2;

    int min_len = sb->len + addition + 1;
    if (sb->cap < min_len) {
        sb->cap = min_len;
    }

    char *buf = milone_mem_alloc(sb->cap, sizeof(char));
    memcpy(buf, sb->buf, sb->len);
    sb->buf = buf;

    assert(sb->len < sb->cap);
    sb->buf[sb->len] = '\0';
}

static void string_builder_append_string(struct StringBuilder *sb,
                                         struct String value) {
    if (sb->len + value.len >= sb->cap) {
        string_builder_grow(sb, value.len);
    }

    memcpy(&sb->buf[sb->len], value.str, value.len + 1);
    sb->len += value.len;
    assert(sb->buf[sb->len] == '\0');
}

int str_cmp(struct String left, struct String right) {
    int min_len = int_clamp(left.len, 0, right.len);
    return memcmp(left.str, right.str, min_len + 1);
}

/// Create a string from a slice of native C string.
struct String str_of_raw_parts(char const *p, int len) {
    if (len <= 0) {
        if (len < 0) {
            fprintf(stderr, "FATAL: Negative string length (%d).\n", len);
            abort();
        }

        return (struct String){.str = "", .len = 0};
    }

    char *str = (char *)milone_mem_alloc(len + 1, sizeof(char));
    memcpy(str, p, len * sizeof(char));
    str[len] = '\0';
    return (struct String){.str = str, .len = len};
}

struct String str_add(struct String left, struct String right) {
    if (left.len == 0 || right.len == 0) {
        return right.len == 0 ? left : right;
    }
    int len = left.len + right.len;
    char *str = (char *)milone_mem_alloc(len + 1, sizeof(char));
    memcpy(str, left.str, left.len);
    memcpy(str + left.len, right.str, right.len);
    assert(str[len] == '\0');
    return (struct String){.str = str, .len = len};
}

struct String str_get_slice(int l, int r, struct String s) {
    l = int_clamp(l, 0, s.len);
    r = int_clamp(r + 1, l, s.len);
    int len = r - l;
    char *str;
    if (r == s.len) {
        str = s.str + l;
    } else {
        str = (char *)milone_mem_alloc(len + 1, sizeof(char));
        memcpy(str, s.str + l, len);
    }
    assert(str[len] == '\0');
    return (struct String){.str = str, .len = len};
}

static void verify_str_to_number(const char *type_name, const char *endptr,
                                 bool range_check) {
    if (!range_check || (*endptr != '\0' && !isspace(*endptr)) ||
        errno == ERANGE) {
        fprintf(stderr, "FATAL: Failed to convert a string to %s.\n",
                type_name);
        exit(1);
    }
}

int8_t str_to_int8(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int8_t", endptr, INT8_MIN <= n && n <= INT8_MAX);
    return (int8_t)n;
}

int16_t str_to_int16(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int16_t", endptr, INT16_MIN <= n && n <= INT16_MAX);
    return (int16_t)n;
}

int str_to_int(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int", endptr, true);
    return n;
}

int64_t str_to_int64(struct String s) {
    char *endptr = s.str + s.len;
    int64_t n = strtoll(s.str, &endptr, 10);
    verify_str_to_number("int64_t", endptr, true);
    return n;
}

intptr_t str_to_intptr(struct String s) {
    char *endptr = s.str + s.len;
    int64_t n = strtoll(s.str, &endptr, 10);
    verify_str_to_number("intptr_t", endptr, true);
    return (intptr_t)n;
}

uint8_t str_to_uint8(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint8_t", endptr, n <= UINT8_MAX);
    return (uint8_t)n;
}

uint16_t str_to_uint16(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint16_t", endptr, n <= UINT16_MAX);
    return (uint16_t)n;
}

uint32_t str_to_uint32(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint32_t", endptr, true);
    return n;
}

uint64_t str_to_uint64(struct String s) {
    char *endptr = s.str + s.len;
    uint64_t n = strtoull(s.str, &endptr, 10);
    verify_str_to_number("uint64_t", endptr, true);
    return n;
}

uintptr_t str_to_uintptr(struct String s) {
    char *endptr = s.str + s.len;
    uint64_t n = strtoull(s.str, &endptr, 10);
    verify_str_to_number("uintptr_t", endptr, true);
    return n;
}

struct String str_of_int64(int64_t value) {
    char buf[21] = {};
    int n = sprintf(buf, "%ld", value);
    return str_of_raw_parts(buf, n);
}

struct String str_of_uint64(uint64_t value) {
    char buf[21] = {};
    int n = sprintf(buf, "%lu", value);
    return str_of_raw_parts(buf, n);
}

double str_to_double(struct String s) {
    char *endptr = s.str + s.len;
    double n = strtod(s.str, &endptr);
    verify_str_to_number("float", endptr, true);
    return n;
}

struct String str_of_double(double value) {
    char buf[64] = {};
    int n = sprintf(buf, "%f", value);
    return str_of_raw_parts(buf, n);
}

struct String str_of_char(char value) {
    char *str = (char *)milone_mem_alloc(2, sizeof(char));
    str[0] = value;
    return (struct String){.str = str, .len = strlen(str)};
}

struct MyStringList {
    struct String head;
    struct MyStringList *tail;
};

struct String str_concat(struct String sep, struct StringList *strings) {
    struct MyStringList *ss = (struct MyStringList *)strings;

    struct StringBuilder *sb = string_builder_new_with_capacity(0x1000);
    bool first = true;

    while (ss) {
        struct String head = ss->head;
        ss = ss->tail;

        if (!first && sep.len > 0) {
            string_builder_append_string(sb, sep);
        }
        first = false;

        string_builder_append_string(sb, head);
    }

    return str_of_raw_parts(sb->buf, sb->len);
}

// -----------------------------------------------
// assertion
// -----------------------------------------------

void milone_assert(bool cond, int y, int x) {
    if (!cond) {
        fprintf(stderr, "Assertion failed at (%d, %d)\n", y + 1, x + 1);
        exit(1);
    }
}

// -----------------------------------------------
// file IO
// -----------------------------------------------

int file_exists(struct String file_name) {
    bool ok = false;

    FILE *fp = fopen(file_name.str, "r");
    if (fp) {
        ok = true;
        fclose(fp);
    }

    return ok;
}

struct String file_read_all_text(struct String file_name) {
    FILE *fp = fopen(file_name.str, "r");
    if (!fp) {
        fprintf(stderr, "File '%s' not found.", file_name.str);
        abort();
    }

    fseek(fp, 0, SEEK_END);
    long size = ftell(fp);
    if (size < 0) {
        fclose(fp);
        fprintf(stderr, "%s", "Couldn't retrieve the file size.");
        abort();
    }
    fseek(fp, 0, SEEK_SET);

    char *content = (char *)milone_mem_alloc((size_t)size + 1, sizeof(char));
    size_t read_size = fread(content, 1, (size_t)size, fp);
    if (read_size != (size_t)size) {
        fclose(fp);
        fprintf(stderr, "%s", "Couldn't retrieve the file contents");
        abort();
    }

    fclose(fp);
    return (struct String){.str = content, .len = size};
}

void file_write_all_text(struct String file_name, struct String content) {
    FILE *fp = fopen(file_name.str, "w");
    if (!fp) {
        fprintf(stderr, "File '%s' not found.", file_name.str);
        abort();
    }

    fprintf(fp, "%s", content.str);

    fclose(fp);
}

struct String milone_get_env(struct String name) {
    char const *value = getenv(name.str);
    if (value == NULL) {
        return (struct String){.str = "", .len = 0};
    }

    return str_of_raw_parts(value, strlen(value));
}

// -----------------------------------------------
// profiling
// -----------------------------------------------

static long milone_get_time_millis(void) {
    struct timespec t;
    timespec_get(&t, TIME_UTC);
    return t.tv_sec * 1000L + t.tv_nsec / (1000L * 1000L);
}

struct Profiler {
    long epoch;
    long heap_size;
};

void *milone_profile_init(void) {
    struct Profiler *p =
        (struct Profiler *)milone_mem_alloc(1, sizeof(struct Profiler));
    p->epoch = milone_get_time_millis();
    p->heap_size = s_heap_size;
    return p;
}

void milone_profile_log(struct String msg, void *profiler) {
    struct Profiler *p = (struct Profiler *)profiler;

    long t = milone_get_time_millis();
    long s = p->epoch;

    long millis = t - s;
    if (millis < 0) {
        millis = 0;
    }

    long sec = millis / 1000;
    millis %= 1000;

    long bytes = s_heap_size - p->heap_size;
    if (bytes < 0) {
        bytes = 0;
    }

    long kilo = bytes / 1000;
    bytes %= 1000;

    long mega = kilo / 1000;
    kilo %= 1000;

    double usage = s_heap_alloc == 0 ? 0.0 : (double)s_heap_size / s_heap_alloc;
    fprintf(stderr, "profile: time=%4d.%03d mem=%5d,%03d,%03d use=%.03f\n%s\n",
            (int)sec, (int)millis, (int)mega, (int)kilo, (int)bytes, usage,
            msg.str);

    p->epoch = t;
    p->heap_size = s_heap_size;
}

// -----------------------------------------------
// For competitive programming
// -----------------------------------------------

int scan_int(void) {
    int value;
    int _n = scanf("%d", &value);
    return value;
}

char scan_char(void) {
    char value;
    int _n = scanf("%c", &value);
    return value;
}

struct String scan_str(int capacity) {
    if (capacity <= 0) {
        fprintf(stderr, "scan_str(%d)", capacity);
        exit(1);
    }

    char *str = calloc(capacity, sizeof(char));
    if (str == NULL) {
        fprintf(stderr, "scan_str(%d) out of memory", capacity);
        exit(1);
    }

    char fmt[16] = {};
    sprintf(fmt, "%%%ds", capacity);
    assert(fmt[15] == 0);

    int _n = scanf(fmt, str);

    int len = strlen(str);
    assert(len < capacity);
    return (struct String){.str = str, .len = len};
}

void *int_array_new(int len) {
    void *p = calloc(len, sizeof(int));
    // fprintf(stderr, "int_array_new(len=%d) = %p\n", len, p);
    return p;
}

int int_array_get(void *array, int index) {
    int value = ((int *)array)[index];
    // fprintf(stderr, "int_array_get(%p, index=%d) = %d\n", array, index,
    // value);
    return value;
}

void int_array_set(void *array, int index, int value) {
    // fprintf(stderr, "int_array_set(%p, index=%d, value=%d)\n", array, index,
    // value);
    ((int *)array)[index] = value;
}

void int_array_copy(void *dest, void *src, int size) {
    memcpy(dest, src, size);
}

// -----------------------------------------------
// Command-line Arguments
// -----------------------------------------------

static int s_argc;
static char **s_argv;

int arg_count(void) { return s_argc; }

struct String arg_get(int index) {
    if (!(0 <= index && index < s_argc)) {
        abort();
    }

    char *str = s_argv[index];
    int len = strlen(str);
    return (struct String){.str = str, .len = len};
}

// -----------------------------------------------
// Runtime Entrypoint
// -----------------------------------------------

#undef main

int main(int argc, char **argv) {
    s_argc = argc - 1;
    s_argv = argv + 1;

    int milone_main();
    return milone_main();
}
