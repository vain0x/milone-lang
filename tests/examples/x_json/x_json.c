#include "milone.h"

struct StringList;

struct StringList* go_(struct StringList* acc_, struct StringList* xs_1);

struct StringList* listRev_(struct StringList* xs_);

int printList_(struct StringList* xs_3);

int isDigit_(char c_);

int readInt_(struct String s_, int i_);

struct StringList* go_1(struct String s_, struct StringList* acc_1, int i_1);

struct StringList* tokenize_(struct String s_);

int main();

struct StringList {
    struct String head;
    struct StringList* tail;
};

struct StringList* go_(struct StringList* acc_, struct StringList* xs_1) {
tailrec_1:;
    struct StringList* match_;
    if ((!((!(xs_1))))) goto next_3;
    match_ = acc_;
    goto end_match_2;
next_3:;
    if ((!(xs_1))) goto next_4;
    struct String x_ = xs_1->head;
    struct StringList* xs_2 = xs_1->tail;
    struct StringList* list_ = (struct StringList*)milone_mem_alloc(1, sizeof(struct StringList));
    list_->head = x_;
    list_->tail = acc_;
    struct StringList* arg_ = list_;
    struct StringList* arg_1 = xs_2;
    acc_ = arg_;
    xs_1 = arg_1;
    goto tailrec_1;
    match_ = NULL;
    goto end_match_2;
next_4:;
    exit(1);
end_match_2:;
    return match_;
}

struct StringList* listRev_(struct StringList* xs_) {
    struct StringList* call_ = go_(NULL, xs_);
    return call_;
}

int printList_(struct StringList* xs_3) {
tailrec_5:;
    int match_1;
    if ((!((!(xs_3))))) goto next_7;
    match_1 = 0;
    goto end_match_6;
next_7:;
    if ((!(xs_3))) goto next_8;
    struct String x_1 = xs_3->head;
    struct StringList* xs_4 = xs_3->tail;
    printf("%s\n", x_1.str);
    int call_1 = 0;
    struct StringList* arg_2 = xs_4;
    xs_3 = arg_2;
    goto tailrec_5;
    match_1 = 0;
    goto end_match_6;
next_8:;
    exit(1);
end_match_6:;
    return 0;
}

int isDigit_(char c_) {
    int match_2;
    if ((c_ < '0')) goto next_10;
    match_2 = ('9' >= c_);
    goto end_match_9;
next_10:;
    if ((c_ >= '0')) goto next_11;
    match_2 = 0;
    goto end_match_9;
next_11:;
    exit(1);
end_match_9:;
    return match_2;
}

int readInt_(struct String s_, int i_) {
tailrec_12:;
    int match_3;
    int match_4;
    if ((i_ < s_.len)) goto next_15;
    match_4 = 1;
    goto end_match_14;
next_15:;
    if ((i_ >= s_.len)) goto next_16;
    int call_2 = isDigit_(s_.str[i_]);
    match_4 = (!(call_2));
    goto end_match_14;
next_16:;
    exit(1);
end_match_14:;
    if ((!(match_4))) goto next_17;
    match_3 = i_;
    goto end_match_13;
next_17:;
    if (match_4) goto next_18;
    struct String arg_3 = s_;
    int arg_4 = (i_ + 1);
    s_ = arg_3;
    i_ = arg_4;
    goto tailrec_12;
    match_3 = 0;
    goto end_match_13;
next_18:;
    exit(1);
end_match_13:;
    return match_3;
}

struct StringList* go_1(struct String s_, struct StringList* acc_1, int i_1) {
tailrec_19:;
    struct StringList* match_5;
    if ((i_1 < s_.len)) goto next_21;
    struct StringList* call_3 = listRev_(acc_1);
    match_5 = call_3;
    goto end_match_20;
next_21:;
    if ((i_1 >= s_.len)) goto next_22;
    struct StringList* match_6;
    if ((s_.str[i_1] != ' ')) goto next_24;
    struct String arg_5 = s_;
    struct StringList* arg_6 = acc_1;
    int arg_7 = (i_1 + 1);
    s_ = arg_5;
    acc_1 = arg_6;
    i_1 = arg_7;
    goto tailrec_19;
    match_6 = NULL;
    goto end_match_23;
next_24:;
    if ((s_.str[i_1] == ' ')) goto next_25;
    struct StringList* match_7;
    int match_8;
    int match_9;
    if ((s_.str[i_1] != '[')) goto next_29;
    match_9 = 1;
    goto end_match_28;
next_29:;
    if ((s_.str[i_1] == '[')) goto next_30;
    match_9 = (s_.str[i_1] == ']');
    goto end_match_28;
next_30:;
    exit(1);
end_match_28:;
    if ((!(match_9))) goto next_31;
    match_8 = 1;
    goto end_match_27;
next_31:;
    if (match_9) goto next_32;
    match_8 = (s_.str[i_1] == ',');
    goto end_match_27;
next_32:;
    exit(1);
end_match_27:;
    if ((!(match_8))) goto next_33;
    struct String call_4 = str_get_slice(i_1, i_1, s_);
    struct StringList* list_1 = (struct StringList*)milone_mem_alloc(1, sizeof(struct StringList));
    list_1->head = str_add((struct String){.str = "punct ", .len = 6}, call_4);
    list_1->tail = acc_1;
    struct StringList* acc_2 = list_1;
    struct String arg_8 = s_;
    struct StringList* arg_9 = acc_2;
    int arg_10 = (i_1 + 1);
    s_ = arg_8;
    acc_1 = arg_9;
    i_1 = arg_10;
    goto tailrec_19;
    match_7 = NULL;
    goto end_match_26;
next_33:;
    if (match_8) goto next_34;
    struct StringList* match_10;
    int call_5 = isDigit_(s_.str[i_1]);
    if ((!(call_5))) goto next_36;
    int call_6 = readInt_(s_, i_1);
    int r_ = call_6;
    struct String call_7 = str_get_slice(i_1, (r_ - 1), s_);
    struct StringList* list_2 = (struct StringList*)milone_mem_alloc(1, sizeof(struct StringList));
    list_2->head = str_add((struct String){.str = "int ", .len = 4}, call_7);
    list_2->tail = acc_1;
    struct StringList* acc_3 = list_2;
    struct String arg_11 = s_;
    struct StringList* arg_12 = acc_3;
    int arg_13 = r_;
    s_ = arg_11;
    acc_1 = arg_12;
    i_1 = arg_13;
    goto tailrec_19;
    match_10 = NULL;
    goto end_match_35;
next_36:;
    if (call_5) goto next_37;
    printf("ERROR: Unknown Token \'%c\'\n", s_.str[i_1]);
    int call_8 = 0;
    exit(1);
    match_10 = NULL;
    goto end_match_35;
next_37:;
    exit(1);
end_match_35:;
    match_7 = match_10;
    goto end_match_26;
next_34:;
    exit(1);
end_match_26:;
    match_6 = match_7;
    goto end_match_23;
next_25:;
    exit(1);
end_match_23:;
    match_5 = match_6;
    goto end_match_20;
next_22:;
    exit(1);
end_match_20:;
    return match_5;
}

struct StringList* tokenize_(struct String s_) {
    struct StringList* call_9 = go_1(s_, NULL, 0);
    return call_9;
}

int main() {
    struct String source_ = (struct String){.str = "[0, 1000000007, 314159]", .len = 23};
    struct StringList* call_10 = tokenize_(source_);
    struct StringList* tokens_ = call_10;
    int call_11 = printList_(tokens_);
    return 0;
}
