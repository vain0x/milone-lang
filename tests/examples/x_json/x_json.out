punct [
int 0
punct ,
int 1000000007
punct ,
int 314159
punct ]
$? = 0
